crates/stark-backend/src/prover/hal.rs

```diff
@@ -0,0 +1,180 @@
+    /// Challenge field (extension field of base field), on host.
+    type Challenge: Copy + Send + Sync + Serialize + DeserializeOwned;
+    /// PCS opening proof on host (see [OpeningProver]). This should not be a reference.
+    type OpeningProof: Clone + Send + Sync + Serialize + DeserializeOwned;
+    /// Single commitment on host.
+    // Commitments are small in size and need to be transferred back to host to be included in proof.
+    type Commitment: Clone + Send + Sync + Serialize + DeserializeOwned;
+    /// Challenger to observe commitments. Sampling is left to other trait implementations.
+    /// We anticipate that the challenger largely operates on the host.
+    type Challenger: CanObserve<Self::Val> + CanObserve<Self::Commitment>;
+
+    // ==== Device Types ====
+    /// Single matrix buffer on device together with dimension metadata.
+    type MatrixView: MatrixView + Clone + Send + Sync;
+    /// Reference for the preimage of a PCS commitment on device, together with any metadata
+    /// necessary for computing opening proofs.
+    ///
+    /// For example, multiple buffers for LDE matrices, their trace domain sizes, and pointer to mixed merkle tree.
+    type PcsDataView: Clone + Send + Sync;
+}
+
+pub trait MatrixView {
+    fn height(&self) -> usize;
+    fn width(&self) -> usize;
+}
+
+pub trait ProverDevice<PB: ProverBackend>:
+    TraceCommitter<PB> + RapPartialProver<PB> + QuotientCommitter<PB> + OpeningProver<PB>
+{
+}
+
+/// Provides functionality for committing to a batch of trace matrices, possibly of different heights.
+pub trait TraceCommitter<PB: ProverBackend> {
+    fn commit(&self, traces: &[PB::MatrixView]) -> (PB::Commitment, PB::PcsDataView);
+}
+
+/// This trait is responsible for all partial proving of after challenge rounds (a.k.a layers) in a
+/// RAP after the main trace has been committed.
+///
+/// The partial prover *may*:
+/// - observe and/or sample challenges
+/// - commit to additional trace data
+/// - generate other partial proof data
+pub trait RapPartialProver<PB: ProverBackend> {
+    // ==== Host Types ====
+    /// Partial proof for multiple RAPs
+    type RapPartialProof: Clone + Send + Sync;
+    /// Part of proving key for a single RAP specific for the RAP challenge phases
+    type RapPartialProvingKeyView<'a>: Clone + Send + Sync
+    where
+        Self: 'a;
+
+    /// The `trace_views` are the views of the respective trace matrices, evaluated on the trace domain.
+    /// Currently this function does not provide a view of any already committed data associated
+    /// with the trace views, although that data is available.
+    fn partially_prove<'a>(
+        &self,
+        challenger: &mut PB::Challenger,
+        pk_views: &[StarkProvingKeyView<'a, PB, Self::RapPartialProvingKeyView<'a>>],
+        trace_views: Vec<PairView<PB::MatrixView, PB::Val>>,
+    ) -> (Self::RapPartialProof, ProverViewAfterRapPhases<PB>);
```

Comment:
I feel it's better to return the after challenge trace matrix to the prover rather than an reference like `PB::MatrixView`.

Reply:
hmm why? I basically never want any large data object (matrix, pcs data, etc) to ever be used on host. the after challenge trace matrix is only used for quotient poly, and we plan for quotient poly to be done entirely on device

Reply:
@kunxian-xia regarding non-view `Matrix` and `PcsData`, do you want a type for device pointer but where the pointer is fully owned and there's no risk that deallocating would affect other views of it, or do you really want it to backed by memory on the host? I could see a case for the former, but I didn't want the latter because it doesn't seem like a pattern to promote.

Reply:
Yes, the `PcsData` for GPU backend is almost identical to `DevicePointer`. The rust borrow checker can decide when is immutable or mutable borrow allowed.

Reply:
> we plan for quotient poly to be done entirely on device

It's possible the gpu memory is not enough to hold all matrices and `PcsData` for all chips.

Reply:
OK I see. Let me address it in a follow up PR. I think `PcsData` I basically don't need to have multiple references to. Matrix is different story.

BTW `PcsDataView` is a little more than just traditional `PcsProverData<SC>`: I want it to include ordered list of trace heights within that commit, because you need it for the opening proof.